---
title: "Introduction to MetaScope"
author: 
- name: Aubrey Odom
  affiliation: 
  - Program in Bioinformatics, Boston University, Boston, MA
  email: aodom@bu.edu
- name: Rahul Varki
  affiliation:
  - Research Assistant in Johnson Lab, Boston University School of Medicine, Boston, MA
  email: rvarki@bu.edu
- name: W. Evan Johnson
  affiliation:
  - The Section of Computational Biomedicine, Boston University School of Medicine, Boston, MA
  email: wej@bu.edu
date: '`r format(Sys.Date(), "%B %e, %Y")`'
package: MetaScope
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{"Introduction to MetaScope"}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

# Important Note

Although it is not a requirement, MetaScope will fun faster and more efficiently for larger samples if the samtools package is present on your device. You can download it at <https://github.com/samtools/samtools>.

# Introduction

MetaScope is a complete metagenomics profiling package that can accurately identify the composition of microbes within a sample at a strain-level resolution. MetaScope can be considered as an updated and expanded R translation of [PathoScope 2.0](https://microbiomejournal.biomedcentral.com/articles/10.1186/2049-2618-2-33), a Python-based metagenomic profiling package created by the Johnson lab. A few improvements made in MetaScope include using the BAM file format instead of the SAM file format for significantly less disk space usage, removing all dependencies to NCBI's now defunct GI sequence annotations, and properly filtering reads that align to filter reference genomes. Functions to analyze host microbiome data are also planned to be added in future updates to the package.

## MetaScope Workflows

The major workflow of MetaScope is delineated below. It is composed of core modules that are formed by groups  of functions.

![Figure 1. The MetaScope workflow and its associated modules with function descriptions. The MetaRef, MetaAlign, MetaFilter, and MetaID modules form the backbone of package operation, whereas the MetaDemultiplex, MetaBLAST, and MetaCombine modules are complementary to the core package functionality.](MetaScope_Workflow.png)

The core modules are as follows:
1. **MetaDemultiplex:** Obtain non-barcoded sequencing reads
2. **MetaRef:** Obtain target and filter genome sequences from NCBI nucleotide database and index using a given aligner
3. **MetaAlign:** Align sequencing reads to indexed target genome sequences
4. **MetaFilter:** Remove reads mapped to indexed host genome sequences
5. **MetaID:** Reassign ambiguously mapped reads to probable source genome
6. **MetaBLAST:** BLAST assigned reads against the NCBI nucleotide database to check identity
7. **MetaCombine:** Aggregate samples into a MultiAssayExperiment compatible with the `animalcules` R package.

There are two sub-workflows that are included in the package, as seen in Figure 1: the Rbowtie2 and the Rsubread workflow. The major difference is that the functions in the MetaRef, MetaAlign, and MetaFilter modules differ by the aligner utilized. 

The Rbowtie2 functions utilize the [Bowtie2](https://www.nature.com/articles/nmeth.1923) aligner (Langmead 2012) whereas the [Rsubread](https://academic.oup.com/nar/article/47/8/e47/5345150) functions utilize the Rsubread aligner (Liao 2019). The nuances of how to use each function can be found by looking at each function's help manual (R command: `?<name of function>`). For reference, PathoScope 2.0 uses the Bowtie2 aligner in its workflow.

In this vignette, we will analyze the mock data provided in the package via the Rbowtie2 sub-workflow. We will utilize all of the core modules in sequential order. We will make mention of the equivalent Rsubread function whenever an Rbowtie2 function is being used. For the purposes of this example, MetaBLAST and MetaCombine modules will be omitted.

## Installation

In order to install MetaScope from Bioconductor, run the following code:

```{R show_install, eval = FALSE}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install("MetaScope")
```

```{R load_packages, eval = TRUE}
suppressPackageStartupMessages({
  library(MetaScope)
  library(magrittr)
})
```

# Data

The mock data provided in the package consists of simulated sequencing data generated from the `SAMtools` `wgsim` function (see `extdata_explanations.Rmd` in `inst/script` to see exact commands). The `wgsim` function is a tool which allows for the generation of FASTQ reads from a reference genome (FASTA). The mock data (`reads.fastq`) contains 1500 reads, of which 1000 reads are derived from the *Staphylococcus aureus* RF122 strain and 500 reads are derived from the *Staphylococcus epidermidis* RP62A strain. In this data set, we assume that the *S. aureus* RF122 reads are the reads of interest, and the *S. epidermidis* RP62A reads are known contaminant reads which should be removed during the analysis. Ideally, the microbial composition report (.csv) produced at the end of the analysis should contain only reads assigned to the *S. aureus* RF122 strain.

# MetaDemultiplex: Demultiplexing reads

Sequence runs on NGS instruments are typically carried out with multiple samples pooled together. An index tag (also called a barcode) consisting of a unique sequence of between 6 and 12bp is added to each sample so that the sequence reads from different samples can be identified. For 16s experiments or sequencing conducted using an Illumina machine, the process of demultiplexing (dividing your sequence reads into separate files for each index tag/sample) and generating the FASTQ data files required for downstream analysis can be done using the MetaScope demultiplexing workflow. This consists of the `meta_demultiplex()` function, which takes as arguments a matrix of sample names/barcodes, a FASTQ file of barcodes by sequence header, and a FASTQ file of reads corresponding to the barcodes. Based on the barcodes given, the function extracts all reads for the indexed barcode and writes all the reads from that barcode to separate FASTQ files.

This is an optional step in the analysis process depending on whether your reads are multiplexed. The reads which we are currently trying to analyze are not multiplexed and therefore this step is skipped in our analysis. The example shown below is using different reads that are barcoded in order to show the utility of the function.

```{R meta_demultiplex, message = FALSE}
# Get barcode, index, and read data locations
barcodePath <-
  system.file("extdata", "barcodes.txt", package = "MetaScope")
indexPath <- system.file("extdata", "virus_example_index.fastq",
                         package = "MetaScope")
readPath <-
  system.file("extdata", "virus_example.fastq", package = "MetaScope")

# Get barcode, index, and read data locations
demult <-
  meta_demultiplex(barcodePath,
                   indexPath,
                   readPath,
                   rcBarcodes = FALSE,
                   hammingDist = 2,
                   location = tempfile())
demult
```

# MetaRef: Reference Genome Library

The MetaScope genome library workflow is designed to assist with collection of sequence data from the National Center for Biotechnology Information (NCBI) nucleotide database. Prior to doing so, the potential targets and filters for the analysis should be identified. That is, what "target" species do you expect to find in your metagenomic sample that you would like to identify, and what reads would you like to "filter" out from the data that are not essential to your analysis?

Typically, the targets of the analysis are microbes (that is, viruses, bacteria, and fungi), and we wish to filter out or discard any reads from the host in addition to artificially added sequences, such as Phi X 174. Following identification of the targets and filters, we use a reference genome library to align the vast number of sample reads back to the respective regions of origin in various species.

The `download_refseq()` function automatically extracts custom reference genome libraries in a FASTA file for microbial or host genomes. The user must first indicate a taxon for which to download the genomes, such as 'bacteria' or 'primates'. A table of possible entries can be viewed by accessing the `MetaScope:::taxonomy_table` object. The user may then specify whether they wish to download only the RefSeq reference genomes, or both the reference and representative genomes. The compress option then allows users to specify whether to compress the output FASTA file; this is done by default.

## Creating a Taxonomy Database

A reference of taxonomy reference identifiers (e.g. NCBI accessions) is critical for fully identifying microbes with the MetaScope pipeline. This step should be completed ONCE, PRIOR TO RUNNING ANY AND ALL SAMPLES with the `download_accessions()` function. For an example of how to make a typical database for all host and filter genomes in your sample, please see the [example script](https://wejlab.github.io/metascope-docs/articles/docs/Example_Script.html) on our website.

If using the SILVA 16S database, NCBI nucleotide or RefSeq database, or NCBI BLAST 16S database, you can use `download_accessions()` with the following arguments:

```{R eval = FALSE}
download_accessions(
  ind_dir = "C:/Users/JohnSmith/ResearchIndices",
  # For NCBI RefSeq or nucleotide database
  NCBI_accessions_database = TRUE,
  NCBI_accessions_name = "accessionTaxa.sql",
  # For SILVA 16S database
  silva_taxonomy_database = TRUE,
  silva_taxonomy_name = "all_silva_headers.rds",
  # For BLAST 16S database
  blast_16S_database = TRUE,
  blast_16S_name = "16S_ribosomal_RNA"
)
```

However, for this vignette, we know in advance precisely which genomes and strains we will be testing. Therefore, we will generate a small subdatabase of taxonomizr references. We do not need to use `download_accessions()` in this case.

```{R taxonomy_db, eval = TRUE, warning = FALSE, message = FALSE}
namesText<-c(
 "1\t|\troot\t|\t\t|\tscientific name\t|",
 "2\t|\tBacteria\t|\tBacteria <prokaryotes>\t|\tscientific name\t|",
 "158878\t|\tStaphylococcus aureus subsp. aureus Mu50\t|\t|\tscientific name\t|",
 "418127\t|\tStaphylococcus aureus subsp. aureus Mu3\t|\t|\tscientific name\t|",
 "426430\t|\tStaphylococcus aureus subsp. aureus str. Newman\t|\t|\tscientific name\t|",
 "158879\t|\tStaphylococcus aureus subsp. aureus N315\t|\t|\tscientific name\t|",
 "273036\t|\tStaphylococcus aureus RF122\t|\t|\tscientific name\t|",
 "523796\t|\tStaphylococcus aureus subsp. aureus ST398\t|\t|\tscientific name\t|",
 "176279\t|\tStaphylococcus epidermidis RP62A\t|\t|\tscientific name\t|",
 "1280\t|\tStaphylococcus aureus\t|\t|\tscientific name\t|",
 "1282\t|\tStaphylococcus epidermidis\t|\t|\tscientific name\t|",
 "1279\t|\tStaphylococcus\t|\t|\tscientific name\t|",
 "90964\t|\tStaphylococcaceae\t|\t|\tscientific name\t|", 
 "1385\t|\tBacillales\t|\t|\tscientific name\t|",
 "91061\t|\tBacilli\t|\t|\tscientific name\t|", 
 "1239\t|\tBacillota\t|\t|\tscientific name\t|",
 "1783272\t|\tBacillati\t|\t|\tscientific name\t|"
)

nodesText<-c(
  "1\t|\t1\t|\tno rank\t|\t\t|\t8\t|\t0\t|\t1\t|\t0\t|\t0\t|\t0\t|\t0\t|\t0\t|\t\t|",
  "2\t|\t131567\t|\tsuperkingdom\t|\t\t|\t0\t|\t0\t|\t11\t|\t0\t|\t0\t|\t0\t|\t0\t|\t0\t|\t\t|",
  "158878\t|\t1280\t|\tstrain",
  "418127\t|\t1280\t|\tstrain",
  "426430\t|\t1280\t|\tstrain",
  "158879\t|\t1280\t|\tstrain",
  "273036\t|\t1280\t|\tstrain",
  "523796\t|\t1280\t|\tstrain",
  "176279\t|\t1282\t|\tstrain",
  "1280\t|\t1279\t|\tspecies",
  "1282\t|\t1279\t|\tspecies",
  "1279\t|\t90964\t|\tgenus",
  "90964\t|\t1385\t|\tfamily", 
  "1385\t|\t91061\t|\torder",
  "91061\t|\t1239\t|\tclass", 
  "1239\t|\t1783272\t|\tphylum",
  "1783272\t|\t2\t|\tkingdom",
  "131567\t|\t1\t|\tno rank"
)

taxa <- c(
  "NA\taccession\ttaxid\tNA",
  "NA\tNC_007622.1\t273036\tNA",
  "NA\tNC_009782.1\t418127\tNA",
  "NA\tNZ_CTYB01000001.1\t418127\tNA",
  "NA\tNZ_CTYB01000002.1\t418127\tNA",
  "NA\tNZ_CTYB01000003.1\t418127\tNA",
  "NA\tNZ_CTYB01000004.1\t418127\tNA",
  "NA\tNZ_CTYB01000005.1\t418127\tNA",
  "NA\tNZ_CTYB01000006.1\t418127\tNA",
  "NA\tNZ_CTYB01000007.1\t418127\tNA",
  "NA\tNZ_CTYB01000008.1\t418127\tNA",
  "NA\tNZ_CTYB01000009.1\t418127\tNA",
  "NA\tNZ_CTYB01000010.1\t418127\tNA",
  "NA\tNZ_CTYB01000011.1\t418127\tNA",
  "NA\tNZ_CTYB01000012.1\t418127\tNA",
  "NA\tNZ_CTYB01000013.1\t418127\tNA",
  "NA\tNZ_CTYB01000014.1\t418127\tNA",
  "NA\tNZ_CTYB01000015.1\t418127\tNA",
  "NA\tNZ_CTYB01000016.1\t418127\tNA",
  "NA\tNZ_CTYB01000017.1\t418127\tNA",
  "NA\tNZ_CTYB01000018.1\t418127\tNA",
  "NA\tNZ_CTYB01000019.1\t418127\tNA",
  "NA\tNZ_CTYB01000020.1\t418127\tNA",
  "NA\tNZ_CTYB01000021.1\t418127\tNA",
  "NA\tNZ_CTYB01000022.1\t418127\tNA",
  "NA\tNZ_CTYB01000023.1\t418127\tNA",
  "NA\tNZ_CTYB01000024.1\t418127\tNA",
  "NA\tNZ_CTYB01000025.1\t418127\tNA",
  "NA\tNZ_CTYB01000026.1\t418127\tNA",
  "NA\tNZ_CTYB01000027.1\t418127\tNA",
  "NA\tNZ_CTYB01000028.1\t418127\tNA",
  "NA\tNZ_CTYB01000029.1\t418127\tNA",
  "NA\tNZ_CTYB01000030.1\t418127\tNA",
  "NA\tNZ_CTYB01000031.1\t418127\tNA",
  "NA\tNZ_CTYB01000032.1\t418127\tNA",
  "NA\tNZ_CTYB01000033.1\t418127\tNA",
  "NA\tNZ_CTYB01000034.1\t418127\tNA",
  "NA\tNZ_CTYB01000035.1\t418127\tNA",
  "NA\tNZ_CTYB01000037.1\t418127\tNA",
  "NA\tNZ_CTYB01000038.1\t418127\tNA",
  "NA\tNZ_CTYB01000039.1\t418127\tNA",
  "NA\tNZ_CTYB01000040.1\t418127\tNA",
  "NA\tNZ_CTYB01000041.1\t418127\tNA",
  "NA\tNZ_CTYB01000042.1\t418127\tNA",
  "NA\tNZ_CTYB01000043.1\t418127\tNA",
  "NA\tNZ_CTYB01000044.1\t418127\tNA",
  "NA\tNZ_CTYB01000045.1\t418127\tNA",
  "NA\tNZ_CTYB01000047.1\t418127\tNA",
  "NA\tNZ_CTYB01000048.1\t418127\tNA",
  "NA\tNZ_CTYB01000049.1\t418127\tNA",
  "NA\tNZ_CTYB01000051.1\t418127\tNA",
  "NA\tNZ_CTYB01000052.1\t418127\tNA",
  "NA\tNZ_CTYB01000053.1\t418127\tNA",
  "NA\tNZ_CTYB01000054.1\t418127\tNA",
  "NA\tNZ_CTYB01000055.1\t418127\tNA",
  "NA\tNZ_CTYB01000056.1\t418127\tNA",
  "NA\tNZ_CTYB01000057.1\t418127\tNA",
  "NA\tNZ_CTYB01000058.1\t418127\tNA",
  "NA\tNZ_CTYB01000059.1\t418127\tNA",
  "NA\tNZ_CTYB01000060.1\t418127\tNA",
  "NA\tNZ_CTYB01000062.1\t418127\tNA",
  "NA\tNZ_CTYB01000063.1\t418127\tNA",
  "NA\tNZ_CTYB01000065.1\t418127\tNA",
  "NA\tNZ_CTYB01000066.1\t418127\tNA",
  "NA\tNZ_CTYB01000068.1\t418127\tNA",
  "NA\tNZ_CTYB01000069.1\t418127\tNA",
  "NA\tNZ_CTYB01000070.1\t418127\tNA",
  "NA\tNZ_CTYB01000071.1\t418127\tNA",
  "NA\tNZ_CTYB01000075.1\t418127\tNA",
  "NA\tNZ_CTYB01000077.1\t418127\tNA",
  "NA\tNZ_CTYB01000078.1\t418127\tNA",
  "NA\tNZ_CTYB01000079.1\t418127\tNA",
  "NA\tNZ_CTYB01000081.1\t418127\tNA",
  "NA\tNZ_CTYB01000083.1\t418127\tNA",
  "NA\tNZ_CTYB01000085.1\t418127\tNA",
  "NA\tNZ_CTYB01000089.1\t418127\tNA",
  "NA\tNZ_CTYB01000090.1\t418127\tNA",
  "NA\tNZ_CTYB01000096.1\t418127\tNA",
  "NA\tNZ_CTYB01000099.1\t418127\tNA",
  "NA\tNZ_CTYB01000101.1\t418127\tNA",
  "NA\tNZ_CTYB01000110.1\t418127\tNA",
  "NA\tNZ_CTYB01000112.1\t418127\tNA",
  "NA\tNZ_CTYB01000113.1\t418127\tNA",
  "NA\tNZ_CTYB01000114.1\t418127\tNA",
  "NA\tNZ_CTYB01000115.1\t418127\tNA",
  "NA\tNC_002758.2\t158878\tNA",
  "NA\tNC_002745.2\t158878\tNA",
  "NA\tNC_003140.1\t158878\tNA",
  "NA\tNC_017333.1\t523796\tNA",
  "NA\tNC_017334.1\t523796\tNA",
  "NA\tNZ_LAWV01000001.1\t523796\tNA",
  "NA\tNZ_LAWV01000010.1\t523796\tNA",
  "NA\tNZ_LAWV01000011.1\t523796\tNA",
  "NA\tNZ_LAWV01000012.1\t523796\tNA",
  "NA\tNZ_LAWV01000013.1\t523796\tNA",
  "NA\tNZ_LAWV01000014.1\t523796\tNA",
  "NA\tNZ_LAWV01000015.1\t523796\tNA",
  "NA\tNZ_LAWV01000016.1\t523796\tNA",
  "NA\tNZ_LAWV01000017.1\t523796\tNA",
  "NA\tNZ_LAWV01000018.1\t523796\tNA",
  "NA\tNZ_LAWV01000019.1\t523796\tNA",
  "NA\tNZ_LAWV01000002.1\t523796\tNA",
  "NA\tNZ_LAWV01000020.1\t523796\tNA",
  "NA\tNZ_LAWV01000021.1\t523796\tNA",
  "NA\tNZ_LAWV01000022.1\t523796\tNA",
  "NA\tNZ_LAWV01000023.1\t523796\tNA",
  "NA\tNZ_LAWV01000024.1\t523796\tNA",
  "NA\tNZ_LAWV01000025.1\t523796\tNA",
  "NA\tNZ_LAWV01000003.1\t523796\tNA",
  "NA\tNZ_LAWV01000030.1\t523796\tNA",
  "NA\tNZ_LAWV01000004.1\t523796\tNA",
  "NA\tNZ_LAWV01000042.1\t523796\tNA",
  "NA\tNZ_LAWV01000043.1\t523796\tNA",
  "NA\tNZ_LAWV01000005.1\t523796\tNA",
  "NA\tNZ_LAWV01000006.1\t523796\tNA",
  "NA\tNZ_LAWV01000007.1\t523796\tNA",
  "NA\tNZ_LAWV01000008.1\t523796\tNA",
  "NA\tNZ_LAWV01000009.1\t523796\tNA",
  "NA\tNZ_CP021178.1\t523796\tNA",
  "NA\tNC_009641.1\t426430\tNA",
  "NA\tNZ_CP023390.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000001.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000002.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000003.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000004.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000005.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000006.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000007.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000008.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000009.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000010.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000011.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000012.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000013.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000014.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000015.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000016.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000017.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000018.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000020.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000021.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000022.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000028.1\t426430\tNA",
  "NA\tNZ_JBBIAE010000032.1\t426430\tNA"
)

inFile <- tempfile()
writeLines(taxa,inFile)
tmp_accession<-tempfile()
taxonomizr::read.names.sql(textConnection(namesText),tmp_accession, overwrite = TRUE)
taxonomizr::read.nodes.sql(textConnection(nodesText),tmp_accession, overwrite = TRUE)
taxonomizr::read.accession2taxid(inFile,tmp_accession,vocal=TRUE, overwrite = TRUE)
```

## Downloading target genomes

Even though in this scenario we know exactly from where the reads in the mock data (`reads.fastq`) originate, in most cases we may only have a general idea of read origins. Therefore, in the following code, we will download the genome of the *Staphylococcus aureus* RF122 strain along with the genomes of a few other closely related *Staphylococcus aureus* strains from the NCBI RefSeq database. These genomes together will act as our target genome library.


```{R target_lib, eval = TRUE, warning = FALSE, message = FALSE}
target_ref_temp <- tempfile()
dir.create(target_ref_temp)

all_species <- c("Staphylococcus aureus subsp. aureus Mu50",
                 "Staphylococcus aureus subsp. aureus Mu3",
                 "Staphylococcus aureus subsp. aureus str. Newman",
                 "Staphylococcus aureus subsp. aureus N315",
                 "Staphylococcus aureus RF122", 
                 "Staphylococcus aureus subsp. aureus ST398")
sapply(all_species, download_refseq, 
       reference = FALSE, representative = FALSE, compress = TRUE,
       out_dir = target_ref_temp, caching = TRUE, accession_path = tmp_accession)
```

## Downloading filter genomes

We will also download the reference genome and related sequences* of the *Staphylococcus epidermidis* RP62A strain from the NCBI nucleotide database, in an uncompressed FASTA format. This genome will act as our filter library.

* since we are downloading from the nucleotide database with representative and reference = FALSE, several sequences will be downloaded in addition to the main genome.
```{R filter_lib, warning = FALSE, message = FALSE}
filter_ref_temp <- tempfile()
dir.create(filter_ref_temp)

download_refseq(
  taxon = "Staphylococcus epidermidis RP62A",
  representative = FALSE, reference = FALSE,
  compress = TRUE, out_dir = filter_ref_temp,
  caching = TRUE,
  accession_path = tmp_accession)
```

## Creating indices using a given aligner

We now use `mk_bowtie_index()`, a wrapper for the `Rbowtie2::bowtie2_build` function, to generate Bowtie2 compatible indexes from the reference genomes that were previously downloaded. The target and reference genome files (.fasta or .fasta.gz extension) must be placed into their own separate empty directories prior to using the function. This is due to the fact that the function will attempt to build the indexes from all the files present in the directory. The function will give an error if other files (other than .fasta or .fasta.gz) are present in the directory. Depending on the combined size of the reference genomes, the function will automatically create either small (.bt2) or large (.bt2l) Bowtie2 indexes.

The target and filter reference genomes downloaded in the previous step have been combined and renamed to target.fasta and filter.fasta respectively for convenience. These are the files from which the Bowtie2 indexes will be made from.

```{R make_indexes, eval = TRUE}
# Create temp directory to store the Bowtie2 indices
index_temp <- tempfile()
dir.create(index_temp)

# Create target index
mk_bowtie_index(
  ref_dir = target_ref_temp,
  lib_dir = index_temp,
  lib_name = "target",
  overwrite = TRUE
)

# Create filter index
mk_bowtie_index(
  ref_dir = filter_ref_temp,
  lib_dir = index_temp,
  lib_name = "filter",
  overwrite = TRUE
)
```

# Alignment with Reference Libraries

After acquiring the target and filter genome libraries, we then take the sequencing reads from our sample and map them first to the target library and then to the filter library. MetaScope's Rbowtie2 mapping function utilizes the [Bowtie2](https://www.nature.com/articles/nmeth.1923) aligner (Langmead 2012) which maps reads to a reference genome using a full-text minute index based approach. Essentially, the algorithm extracts substrings which are referred to as "seeds" from the reads and aligns them to the reference genomes with the assistance from the full-text minute index. Seed alignments to the reference genomes are prioritized and then finally extended into full alignments using dynamic programming.

## MetaAlign

Following index creation, we will use the Bowtie2 aligner to map the reads to the target genomes with the `align_target_bowtie()` function (Rsubread equivalent: `align_target()`). The function takes as an input the location of the FASTQ file to align, the directory where the indexes are stored, the names of the indexes to align against, the directory where the BAM file should be written, and the basename of the output BAM file. 

In practice, `align_target_bowtie()` maps reads to each target library separately, removes the unmapped reads from each file, and finally merges and sorts by chromosome the BAM files from each library into a single output file (same with `align_target`). If `SAMtools` is installed on the machine and can be found by the `Sys.which("samtools")` R command, the BAM file will be directly created, otherwise an intermediate SAM file will be created prior to the creation of the BAM file which could potentially create issues if the SAM file is large and there is limited disk space. The default alignment parameters are the same as PathoScope 2.0's default alignment parameters, but users can provide their own Bowtie 2 alignment settings if desired.

We will now align the sample reads (reads.fastq) to the target reference genomes using the Bowtie 2 indexes that we just built.

```{R alignment_align}
# Create a temp directory to store output bam file
output_temp <- tempfile()
dir.create(output_temp)

# Get path to example reads
readPath <-
  system.file("extdata", "reads.fastq", package = "MetaScope")

# Align reads to the target genomes
target_map <-
  align_target_bowtie(
    read1 = readPath,
    lib_dir = index_temp,
    libs = "target",
    align_dir = output_temp,
    align_file = "bowtie_target",
    overwrite = TRUE
  )
```

## MetaFilter
The last step in the mapping workflow is to filter the output BAM file according to the reference genome for the filter/host species. Although we have already filtered out any unmapped reads, which may belong to one or more host species or otherwise, there may still remain some sort of unwelcome contamination in the data from the filter species which we wish to remove. To do this, we employ `filter_host_bowtie()` (Rsubread equivalent: `filter_host()`), which takes as an input the location of the BAM file created from `align_target_bowtie()`, the directory where the indexes are stored, and the names of the filter indexes to align against, to produce a sorted BAM file with any reads that match the filter libraries removed. We will then use this final BAM file downstream for further analysis.

```{R alignment_filter}
final_map <-
  filter_host_bowtie(
    reads_bam = target_map,
    lib_dir = index_temp,
    libs = "filter",
    make_bam = TRUE, # Set to true to create BAM output
    # Default is to create simplified .csv.gz output
    # The .csv.gz output is much quicker to create!
    overwrite = TRUE,
    threads = 1
  )
```

## Evaluating alignments (supplemental)
**Note: the next two code blocks are included for the sake of examining the vignette example, but are not useful for "real life" data. To continue your analysis, head down to the Genome Identification header.**

Prior to the last step in the analysis, we will look at the primary alignments of the mapped reads in the filtered BAM file that we just created using the `filter_host_bowtie()` function. According to the [Bowtie2 manual](http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml), a primary alignment is described as the alignment that received the highest alignment score among all alignments for that read. When looking at the primary alignments of the mapped reads, we can see that the majority of reads have mapped to the correct *Staphylococcus aureus* RF122 strain. However, some residual reads have primary alignments to the other *S. aureus* strains which are incorrect. If we were to stop the analysis at this point, we could potentially be lead to believe that our sample has increased microbial diversity, when it actually does not.

```{R bam_primary_alignment}
bamFile <- Rsamtools::BamFile(final_map)

param <-
  Rsamtools::ScanBamParam(
    flag = Rsamtools::scanBamFlag(isSecondaryAlignment = FALSE),
    what = c("flag", "rname")
  ) #Gets info about primary alignments

aln <- Rsamtools::scanBam(bamFile, param = param)
accession_all <- aln[[1]]$rname
genome_name_all <- accession_all |> 
  taxonomizr::accessionToTaxa(tmp_accession) |>
  taxonomizr::getTaxonomy(sqlFile = tmp_accession, desiredTaxa = "strain")
read_count_table <- sort(table(genome_name_all), decreasing = TRUE)
knitr::kable(
  read_count_table,
  col.names = c("Genome Assigned", "Read Count"))
```

We can also look at the secondary alignments of the mapped reads within our filtered BAM file. A secondary alignment occurs when a read maps to multiple different genomes. We can see that the majority of our secondary alignments are to the other Staphylococcus aureus strains, which makes sense considering that the majority of the primary alignments were to the correct Staphylococcus aureus RF122 strain.

```{R bam_secondary_alignment}
bamFile <- Rsamtools::BamFile(final_map)

param <-
  Rsamtools::ScanBamParam(
    flag = Rsamtools::scanBamFlag(isSecondaryAlignment = TRUE),
    what = c("flag", "rname")
  ) #Gets info about secondary alignments

aln <- Rsamtools::scanBam(bamFile, param = param)
accession_all <- aln[[1]]$rname
genome_name_all <- accession_all |> 
  taxonomizr::accessionToTaxa(tmp_accession) |>
  taxonomizr::getTaxonomy(sqlFile = tmp_accession, desiredTaxa = "strain")
read_count_table <- sort(table(genome_name_all), decreasing = TRUE)
knitr::kable(
  read_count_table,
  col.names = c("Genome Assigned", "Read Count"))
```

# MetaID: Origin Genome Identification

Following the proper alignment of a sample to all target and filter libraries of interest, we may proceed in identifying which genomes are most likely to be represented in the sample. This identification workflow is the core of MetaScope; it features a Bayesian read reassignment model which dramatically improves specificity and sensitivity over other methods (Francis et. al 2013). This is because such a method identifies reads with unique alignments and uses them to guide the reassignment of reads with ambiguous alignments.

The MetaID identification module consists of a single function, `MetaScope_ID()`, which reads in a .bam file, annotates the taxonomy and genome names, reduces the mapping ambiguity using a mixture model, and outputs a .csv file with the results. Currently, it assumes that the genome library/.bam files use NCBI accession names for reference names.

```{R identification, message = FALSE}
output <- metascope_id(
  final_map,
  input_type = "bam",
  # change input_type to "csv.gz" when not creating a BAM
  aligner = "bowtie2",
  num_species_plot = 0,
  accession_path = tmp_accession
)

knitr::kable(output,
             format = "html",
             digits = 2,
             caption = "Table of MetaScope ID results")
```

**Note: the next code block is included for the sake of examining the vignette example, but can be skipped. Your results are in the CSV file produced by the `metascope_id()` function.**

We will now look at the read reassignment results reported in the output CSV file.

```{R CSV_summary}
relevant_col <- dirname(final_map) %>%
  file.path("bowtie_target.metascope_id.csv") %>%
  read.csv() %>% dplyr::select(2:4)

relevant_col |>
  dplyr::mutate(
    Genome = stringr::str_replace_all(Genome, ',.*', ''),
    Genome = stringr::str_replace_all(Genome, "(ST398).*", "\\1"),
    Genome = stringr::str_replace_all(Genome, "(N315).*", "\\1"),
    Genome = stringr::str_replace_all(Genome, "(Newman).*", "\\1"),
    Genome = stringr::str_replace_all(Genome, "(Mu3).*", "\\1"),
    Genome = stringr::str_replace_all(Genome, "(RF122).*", "\\1")
  ) |>
  knitr::kable()
unlink(".bowtie2.cerr.txt")
```

We can see that the read reassignment function has reassigned the majority of the ambiguous alignments back to the *Staphylococcus aureus* RF122 strain, the correct strain of origin.

# Session Info

```{r session_info}
sessionInfo()
```
